stdlib/toolchain/CompatibilityPacks/Metadata.cpp

```diff
@@ -0,0 +1,164 @@
+inline llvm::hash_code our_hash_integer_value(uint64_t value) {
+  const char *s = reinterpret_cast<const char *>(&value);
+  const uint64_t a = llvm::hashing::detail::fetch32(s);
+  return llvm::hashing::detail::hash_16_bytes(
+      (a << 3), llvm::hashing::detail::fetch32(s + 4));
+}
+
+static inline llvm::hash_code our_hash_combine(llvm::hash_code seed, llvm::hash_code v) {
+  return seed ^ (v + 0x9e3779b9 + (seed<<6) + (seed>>2));
+}
+
+/// Copy and paste from Metadata.cpp.
+
+namespace {
+
+template<typename PackType>
+class PackCacheEntry {
+public:
+  size_t Count;
+
+  const PackType * const * getElements() const {
+    return reinterpret_cast<const PackType * const *>(this + 1);
+  }
+
+  const PackType ** getElements() {
+    return reinterpret_cast<const PackType **>(this + 1);
+  }
+
+  struct Key {
+    const PackType *const *Data;
+    const size_t Count;
+
+    size_t getCount() const {
+      return Count;
+    }
+
+    const PackType *getElement(size_t index) const {
+      assert(index < Count);
+      return Data[index];
+    }
+
+    friend llvm::hash_code hash_value(const Key &key) {
+      llvm::hash_code hash = 0;
+      for (size_t i = 0; i != key.getCount(); ++i) {
+        hash = our_hash_combine(hash, our_hash_integer_value(
+            reinterpret_cast<uint64_t>(key.getElement(i))));
+      }
+      return hash;
+    }
+  };
+
+  PackCacheEntry(const Key &key);
+
+  intptr_t getKeyIntValueForDump() {
+    return 0; // No single meaningful value here.
+  }
+
+  bool matchesKey(const Key &key) const {
+    if (key.getCount() != Count)
+      return false;
+    for (unsigned i = 0; i != Count; ++i) {
+      if (key.getElement(i) != getElements()[i])
+        return false;
+    }
+    return true;
+  }
+
+  friend llvm::hash_code hash_value(const PackCacheEntry<PackType> &value) {
+    llvm::hash_code hash = 0;
+    for (size_t i = 0; i != value.Count; ++i) {
+      hash = our_hash_combine(hash, our_hash_integer_value(
+          reinterpret_cast<uint64_t>(value.getElements()[i])));
+    }
+    return hash;
+  }
+
+  static size_t getExtraAllocationSize(const Key &key) {
+    return getExtraAllocationSize(key.Count);
+  }
+
+  size_t getExtraAllocationSize() const {
+    return getExtraAllocationSize(Count);
+  }
+
+  static size_t getExtraAllocationSize(unsigned count) {
+    return count * sizeof(const Metadata * const *);
+  }
+};
+
+template<typename PackType>
+PackCacheEntry<PackType>::PackCacheEntry(
+    const typename PackCacheEntry<PackType>::Key &key) {
+  Count = key.getCount();
+
+  for (unsigned i = 0; i < Count; ++i)
+    getElements()[i] = key.getElement(i);
+}
+
+} // end anonymous namespace
+
+/// The uniquing structure for metadata packs.
+static SimpleGlobalCache<PackCacheEntry<Metadata>,
+                         MetadataPackTag> MetadataPacks;
+
+SWIFT_RUNTIME_EXPORT SWIFT_CC(swift)
+const Metadata * const *
+swift_allocateMetadataPack(const Metadata * const *ptr, size_t count) {
```

Comment:
I don't think the uniquing here will work correctly when the Swift runtime in use has `swift_allocateMetadataPack` (which uses a runtime-internal `MetadataPacks` cache) and the program was compiled for an earlier deployment target (so this version of `swift_allocateMetadataPack` will be linked in), because some callers will go through the runtime version and some will go through the compatibility version.

With back-deployed concurrency, I ended up creating `swift_getFunctionTypeMetadataGlobalActorBackDeploy` in the back-deployment shim. IRGen generates a call to that when building for a back-deployment target, and that function does a `dlsym` to determine whether `swift_getFunctionTypeMetadataGlobalActor` is available in the active Swift runtime... if it is, we use it so everyone agrees on the underlying cache. I think you'll need to take a similar approach here.